File: duplicates-android/app/src/main/java/com/github/duplicates/message/MessageComparator.java
/*
 * Copyright 2016, Moshe Waisberg
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.duplicates.message;

import android.text.format.DateUtils;

import com.github.duplicates.DuplicateComparator;

/**
 * Compare duplicate messages.
 *
 * @author moshe.w
 */
public class MessageComparator extends DuplicateComparator<MessageItem> {

    public static final int TYPE = 0;
    public static final int DATE = 1;
    public static final int DATE_SENT = 2;
    public static final int ADDRESS = 3;
    public static final int PERSON = 4;
    public static final int BODY = 5;
    public static final int SUBJECT = 6;
    public static final int THREAD_ID = 7;
    public static final int STATUS = 8;
    public static final int ERROR_CODE = 9;
    public static final int LOCKED = 10;
    public static final int PROTOCOL = 11;
    public static final int READ = 12;
    public static final int SEEN = 13;

    @Override
    public int compare(MessageItem lhs, MessageItem rhs) {
        int c;

        c = compare(lhs.getType(), rhs.getType());
        if (c != SAME) {
            return c;
        }
        c = compare(lhs.getDateReceived(), rhs.getDateReceived());
        if (c != SAME) {
            return c;
        }
        c = compare(lhs.getDateSent(), rhs.getDateSent());
        if (c != SAME) {
            return c;
        }
        c = compare(lhs.getAddress(), rhs.getAddress());
        if (c != SAME) {
            return c;
        }
        c = compare(lhs.getPerson(), rhs.getPerson());
        if (c != SAME) {
            return c;
        }
        c = compare(lhs.getBody(), rhs.getBody());
        if (c != SAME) {
            return c;
        }
        c = compare(lhs.getSubject(), rhs.getSubject());
        if (c != SAME) {
            return c;
        }
        c = compare(lhs.getThreadId(), rhs.getThreadId());
        if (c != SAME) {
            return c;
        }
        c = compare(lhs.getStatus(), rhs.getStatus());
        if (c != SAME) {
            return c;
        }
        c = compare(lhs.getErrorCode(), rhs.getErrorCode());
        if (c != SAME) {
            return c;
        }
        c = compare(lhs.isLocked(), rhs.isLocked());
        if (c != SAME) {
            return c;
        }
        c = compare(lhs.getProtocol(), rhs.getProtocol());
        if (c != SAME) {
            return c;
        }
        c = compare(lhs.isRead(), rhs.isRead());
        if (c != SAME) {
            return c;
        }
        c = compare(lhs.isSeen(), rhs.isSeen());
        if (c != SAME) {
            return c;
        }

        return super.compare(lhs, rhs);
    }

    @Override
    public boolean[] difference(MessageItem lhs, MessageItem rhs) {
        boolean[] result = new boolean[14];

        result[ADDRESS] = compareIgnoreCase(lhs.getAddress(), rhs.getAddress()) != SAME;
        result[BODY] = compare(lhs.getBody(), rhs.getBody()) != SAME;
        result[DATE] = compareTime(lhs.getDateReceived(), rhs.getDateReceived(), DateUtils.SECOND_IN_MILLIS) != SAME;
        result[DATE_SENT] = compareTime(lhs.getDateSent(), rhs.getDateSent(), DateUtils.SECOND_IN_MILLIS) != SAME;
        result[ERROR_CODE] = compare(lhs.getErrorCode(), rhs.getErrorCode()) != SAME;
        result[LOCKED] = compare(lhs.isLocked(), rhs.isLocked()) != SAME;
        result[PERSON] = compare(lhs.getPerson(), rhs.getPerson()) != SAME;
        result[PROTOCOL] = compare(lhs.getProtocol(), rhs.getProtocol()) != SAME;
        result[READ] = compare(lhs.isRead(), rhs.isRead()) != SAME;
        result[SEEN] = compare(lhs.isSeen(), rhs.isSeen()) != SAME;
        result[STATUS] = compare(lhs.getStatus(), rhs.getStatus()) != SAME;
        result[SUBJECT] = compareIgnoreCase(lhs.getSubject(), rhs.getSubject()) != SAME;
        result[THREAD_ID] = compare(lhs.getThreadId(), rhs.getThreadId()) != SAME;
        result[TYPE] = compare(lhs.getType(), rhs.getType()) != SAME;

        return result;
    }

    @Override
    public float match(MessageItem lhs, MessageItem rhs, boolean[] difference) {
        if (difference == null) {
            difference = difference(lhs, rhs);
        }
        float match = MATCH_SAME;

        if (difference[DATE]) {
            match *= 0.7f;
        }

        if (difference[TYPE]) {
            match *= 0.8f;
        }
        if (difference[DATE_SENT]) {
            match *= 0.8f;
        }
        if (difference[ADDRESS]) {
            match *= matchTitle(lhs.getAddress(), rhs.getAddress(), 0.8f);
        }
        if (difference[PERSON]) {
            match *= 0.8f;
        }
        if (difference[BODY]) {
            match *= 0.75f;
        }

        if (difference[SUBJECT]) {
            match *= matchTitle(lhs.getSubject(), rhs.getSubject(), 0.85f);
        }
        if (difference[THREAD_ID]) {
            match *= 0.9f;
        }

        if (difference[STATUS]) {
            match *= 0.95f;
        }
        if (difference[ERROR_CODE]) {
            match *= 0.95f;
        }
        if (difference[LOCKED]) {
            match *= 0.95f;
        }
        if (difference[PROTOCOL]) {
            match *= 0.95f;
        }
        if (difference[READ]) {
            match *= 0.95f;
        }
        if (difference[SEEN]) {
            match *= 0.95f;
        }

        return match;
    }
}

File: duplicates-android/app/src/main/java/com/github/duplicates/message/MessageComparator.kt
/*
 * Copyright 2016, Moshe Waisberg
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.duplicates.message

import android.text.format.DateUtils.SECOND_IN_MILLIS
import com.github.duplicates.DuplicateComparator

/**
 * Compare duplicate messages.
 *
 * @author moshe.w
 */
class MessageComparator : DuplicateComparator<MessageItem>() {

    override fun compare(lhs: MessageItem, rhs: MessageItem): Int {
        var c: Int

        c = compare(lhs.type, rhs.type)
        if (c != SAME) return c
        c = compare(lhs.dateReceived, rhs.dateReceived)
        if (c != SAME) return c
        c = compare(lhs.dateSent, rhs.dateSent)
        if (c != SAME) return c
        c = compare(lhs.address, rhs.address)
        if (c != SAME) return c
        c = compare(lhs.person, rhs.person)
        if (c != SAME) return c
        c = compare(lhs.body, rhs.body)
        if (c != SAME) return c
        c = compare(lhs.subject, rhs.subject)
        if (c != SAME) return c
        c = compare(lhs.threadId, rhs.threadId)
        if (c != SAME) return c
        c = compare(lhs.status, rhs.status)
        if (c != SAME) return c
        c = compare(lhs.errorCode, rhs.errorCode)
        if (c != SAME) return c
        c = compare(lhs.isLocked, rhs.isLocked)
        if (c != SAME) return c
        c = compare(lhs.protocol, rhs.protocol)
        if (c != SAME) return c
        c = compare(lhs.isRead, rhs.isRead)
        if (c != SAME) return c
        c = compare(lhs.isSeen, rhs.isSeen)
        return if (c != SAME) c else super.compare(lhs, rhs)
    }

    override fun difference(lhs: MessageItem, rhs: MessageItem): BooleanArray {
        val result = BooleanArray(14)

        result[ADDRESS] = isDifferentIgnoreCase(lhs.address, rhs.address)
        result[BODY] = isDifferent(lhs.body, rhs.body)
        result[DATE] = isDifferentTime(lhs.dateReceived, rhs.dateReceived, SECOND_IN_MILLIS)
        result[DATE_SENT] = isDifferentTime(lhs.dateSent, rhs.dateSent, SECOND_IN_MILLIS)
        result[ERROR_CODE] = isDifferent(lhs.errorCode, rhs.errorCode)
        result[LOCKED] = isDifferent(lhs.isLocked, rhs.isLocked)
        result[PERSON] = isDifferent(lhs.person, rhs.person)
        result[PROTOCOL] = isDifferent(lhs.protocol, rhs.protocol)
        result[READ] = isDifferent(lhs.isRead, rhs.isRead)
        result[SEEN] = isDifferent(lhs.isSeen, rhs.isSeen)
        result[STATUS] = isDifferent(lhs.status, rhs.status)
        result[SUBJECT] = isDifferentIgnoreCase(lhs.subject, rhs.subject)
        result[THREAD_ID] = isDifferent(lhs.threadId, rhs.threadId)
        result[TYPE] = isDifferent(lhs.type, rhs.type)

        return result
    }

    override fun match(lhs: MessageItem, rhs: MessageItem, difference: BooleanArray?): Float {
        val different = difference ?: difference(lhs, rhs)
        var match = MATCH_SAME

        if (different[DATE]) {
            match *= 0.7f
        }
        if (different[TYPE]) {
            match *= 0.8f
        }
        if (different[DATE_SENT]) {
            match *= 0.8f
        }
        if (different[ADDRESS]) {
            match *= matchTitle(lhs.address, rhs.address, 0.8f)
        }
        if (different[PERSON]) {
            match *= 0.8f
        }
        if (different[BODY]) {
            match *= 0.75f
        }
        if (different[SUBJECT]) {
            match *= matchTitle(lhs.subject, rhs.subject, 0.85f)
        }
        if (different[THREAD_ID]) {
            match *= 0.9f
        }
        if (different[STATUS]) {
            match *= 0.95f
        }
        if (different[ERROR_CODE]) {
            match *= 0.95f
        }
        if (different[LOCKED]) {
            match *= 0.95f
        }
        if (different[PROTOCOL]) {
            match *= 0.95f
        }
        if (different[READ]) {
            match *= 0.95f
        }
        if (different[SEEN]) {
            match *= 0.95f
        }

        return match
    }

    companion object {

        val TYPE = 0
        val DATE = 1
        val DATE_SENT = 2
        val ADDRESS = 3
        val PERSON = 4
        val BODY = 5
        val SUBJECT = 6
        val THREAD_ID = 7
        val STATUS = 8
        val ERROR_CODE = 9
        val LOCKED = 10
        val PROTOCOL = 11
        val READ = 12
        val SEEN = 13
    }
}
